### DISCUSSION TITLE
Integrate with OpenWebUI

### DISCUSSION THREAD

**Discussion by mdlmarkham** [ItzCrazyKns/Vane#360] [Ideas]
[OpenWebUI](https://github.com/open-webui) is a great UI for general chatting with local LLMs. I think that Perplexica is very complimentary. What do you think about an integration?

**Comment by Trapper4888** [comment]
You could make a tool/function in openwebui that uses perplexica's api, and have both openwebui and perplexica use your local LLM serving (ollama, tabby, ...).

**Comment by JoeLong54** [comment]
I have modified the docker-compose.yaml file to include open-webui and pipelines it appears to be working, don't know how it will handle updates.
[docker-compose.yaml.txt](https://github.com/user-attachments/files/17462113/docker-compose.yaml.txt)

**Reply by delfireinoso** [reply]
I understand that you have merged all services on a compose file, but you anly need to address the port conflicts between OpenWebUi and Perplexica, and you can put local LAN ip optionally in the Perplexica Compse file to run them both

I install and updathe them on their own compose file so it's easy to update

What I want to see your Pileplines py.file to use Perplexica API

**Reply by yieme** [reply]
Yes, the port conflict are a problem.

**Reply by ViffyGwaanl** [reply]
You can use this : https://openwebui.com/f/gwaanl/perplexica_pipe

**Comment by ViffyGwaanl** [comment]
You can use this : https://openwebui.com/f/gwaanl/perplexica_pipe

**Reply by delfireinoso** [reply]
I’ll install as soon as I finish my winter Holidays. Thanks for your reply. You have modified the YAML from Perplexica? I use Perplexica on 3080:3000 to avoid problems with OpenWeUi

**Reply by ViffyGwaanl** [reply]
> I’ll install as soon as I finish my winter Holidays. Thanks for your reply. You have modified the YAML from Perplexica? I use Perplexica on 3080:3000 to avoid problems with OpenWeUi

![微信截图_20250103093517](https://github.com/user-attachments/assets/f52e757a-5354-423f-abe0-43fdf8898167)
![微信截图_20250103093503](https://github.com/user-attachments/assets/c14c95b0-a1bb-40dd-9aff-8d79cdb98ef4)


I haven't modified the YAML of Perplexica, but you can modify the configuration here

**Reply by delfireinoso** [reply]
I've installed the pipe. I'have Perplexica on the same Docker that OpenWebui, and I'have Perplexica on 

`http://127.0.0.1:3001/api/search `

I only have changed the public port from 3000 to 3080 in Perplexica to have both services accessible by the browser. I use local models with Ollama, that's running out of docker

What I get is:

> An error occurred while searching Perplexica: HTTPConnectionPool(host='127.0.0.1', port=3001): Max retries exceeded with url: /api/search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffff21b977d0>: Failed to establish a new connection: [Errno 111] Connection refused'))

If you have some clues how to fix this I would be glad to try them

**Reply by ViffyGwaanl** [reply]
> I've installed the pipe. I'have Perplexica on the same Docker that OpenWebui, and I'have Perplexica on 
> 
> `http://127.0.0.1:3001/api/search `
> 
> I only have changed the public port from 3000 to 3080 in Perplexica to have both services accessible by the browser. I use local models with Ollama, that's running out of docker 
> 
> What I get is: 
> 
> > An error occurred while searching Perplexica: HTTPConnectionPool(host='127.0.0.1', port=3001): Max retries exceeded with url: /api/search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffff21b977d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
> 
> If you have some clues how to fix this I would be glad to try them

I have also encountered this problem, and you need to directly use the specific IP address of the device in Docker, for example, the "http://192.168.50.116:3001/api/search" as shown in my picture. You can find it in the network settings.

**Reply by ViffyGwaanl** [reply]
> I've installed the pipe. I'have Perplexica on the same Docker that OpenWebui, and I'have Perplexica on
> 
> `http://127.0.0.1:3001/api/search `
> 
> I only have changed the public port from 3000 to 3080 in Perplexica to have both services accessible by the browser. I use local models with Ollama, that's running out of docker
> 
> What I get is:
> 
> > An error occurred while searching Perplexica: HTTPConnectionPool(host='127.0.0.1', port=3001): Max retries exceeded with url: /api/search (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffff21b977d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
> 
> If you have some clues how to fix this I would be glad to try them

I updated the code to support custom model links (ollama, custom_openai, etc.), you can update it：https://github.com/ViffyGwaanl/perplexica-pipe

**Comment by delfireinoso** [comment]
I've installed your new version, I put the IP, ollama and the installed models, a soon as I've made this OpenWebUi has been frozen. so something is wrong. I'll try to figure how to uninstall it. I'm use a Mac Mini Pro with  the latest version of MacOs installed.

I'll ask on the OpenWebUi how to unistall the pipe, if you have a suggestion, i'm all ears

**Comment by delfireinoso** [comment]
[open-webui Docker.log](https://github.com/user-attachments/files/18353883/open-webui.Docker.log)

I have attached the Docker log. If you know how to delete de damaged files and how to start from new without deleting all done on OpenWebUi, will be great

**Reply by ViffyGwaanl** [reply]
Are ollama and openwebui installed independently in separate Docker containers on macOS? I've only tested this on Windows before, and I'll try to reproduce it on my Mac later. I'd like to know how you installed openwebui.
 If you specified a data file path during installation, you should be able to find the webui.db file in the _data folder at that path, which contains all the openwebui data.  Could you describe the openwebui error you're encountering?
 Is the entire Docker project unable to start after installing the code?

**Reply by ViffyGwaanl** [reply]
I checked the log file and found that the error occurred during the validation process of the Valves class. It seems the customOpenAIBaseURL and customOpenAIKey fields were None instead of valid strings.

**Reply by ViffyGwaanl** [reply]
I have reproduced the problem and am looking for a solution

**Reply by ViffyGwaanl** [reply]
I found the problem. You need to locate the file "/app/backend/open_webui/functions.py" within your docker container.  Then, insert the following code snippet between lines 63 and 64.

After adding the code, save the file and restart the openwebui docker container. Once restarted, refresh the webpage and it should launch successfully. 

You can then go into the administrator settings and delete the pipe file.  Afterward, remove the code you added to "/app/backend/open_webui/functions.py".

I will update the pipe script shortly, or you can use version 0.3.2 for now. I'll let you know when the update is ready.


![42d61c66c966795bac602a6035b7e60](https://github.com/user-attachments/assets/daf77fd3-32a5-4967-a883-ce8a0b3d3f5b)
![4a41afbc84d77ad219fa65255fca84f](https://github.com/user-attachments/assets/140dbdda-b329-42eb-b886-f478e6297f40)

```
        if valves is None:
            valves = {}
        if 'customOpenAIBaseURL' not in valves or not isinstance(valves['customOpenAIBaseURL'], str):
            valves['customOpenAIBaseURL'] = 'https://api.xxxxx.com/v1'  
        if 'customOpenAIKey' not in valves or not isinstance(valves['customOpenAIKey'], str):
            valves['customOpenAIKey'] = 'sk-xxxxxxxxxxx' 
```

**Reply by delfireinoso** [reply]
I'm out now so I'll fix it later 

Yes I use Ollama out of Docker as it's easy to maintain in MacOS

And the problem was with OpenwebUi that was not starting after the installation of yesterday version

Perplexica on the same Docker was working 

I'll use your instructions and try them thanks

I'll wait for the next version of your pipe

**Reply by ViffyGwaanl** [reply]
The new 0.3.4 code has resolved the issue where the field being none prevented openwebui from starting. You can find it on GitHub or update in the openwebui community.

**Reply by delfireinoso** [reply]
I've solved the problem with your guide

**Comment by delfireinoso** [comment]
I've installed the new version and It's working sometimes, but as soon as I try to change the configuration I have problems and I need to restart the OpenWebUi docker image, and sometimes I need to do it twice to get it working. But shoud be some detail in the code and the pipe is working

I'm using the last version of openWebUi and Perplexica

**Reply by ViffyGwaanl** [reply]
Could you please specify which configurations you have changed, and could you attach a parameter diagram?

**Reply by delfireinoso** [reply]
![image](https://github.com/user-attachments/assets/7e8de47f-9610-457c-af1a-7d80f05126b1)

These are my settings 

Now I'm accessing from my phone

Nothing is wrong

**Reply by delfireinoso** [reply]
I think that some things have changed in OpenWebUi from 0,52 to 0,54

It takes longer to start in remote 

But everytime I access to settings I have problems in OpenWebUi 

So I'll need to restart the Docker Image when I arrive at home

**Reply by ViffyGwaanl** [reply]
Is this problem happening every time you open it remotely from the web? Will deleting the perplexica pipe improve the situation?

**Reply by ViffyGwaanl** [reply]
I checked the logs and version update content of openwebui and found that after version 0.5.0, the function and pipeline features are no longer independent .py format scripts that are loaded separately after startup, but are merged into the backend to be checked and loaded during system startup. Therefore, it will only start successfully after connecting the pipe. The purpose should be to improve the issue in previous versions where the model list had not loaded after startup. As a result, adding a pipe may slow down the startup speed. I optimized some code in the perplexica pipe script of version 0.3.5 in an attempt to improve the issue. You can try to see if there are any changes. It is recommended to note the configuration, delete the original script, and reinstall.

**Reply by ViffyGwaanl** [reply]
I think the reason why it opens slowly might be that OpenWebUI has to wait for the pipe script to successfully wait for Perplexica to finish loading the local large model from Ollama before it can launch successfully. I have been using online APIs for testing before.
 I will use Ollama for a new round of testing later to find out if there is any reason for this.

**Reply by delfireinoso** [reply]
Yes if I unmarck the pipe OpenWebUi works as expected

So sure is the problem you have noticed: something goes to wait the model upload and something lacks and is waiting more than the time needed

If you can try on your system nice

I'm out for some days

**Reply by delfireinoso** [reply]
![image](https://github.com/user-attachments/assets/4325f412-2845-4107-9301-538626f5944c)

I understand that we need to solve things on the OpenWebUi side and perhaps on the Perplexica API (my models load quickly the problem is on the communication)

In the meantime I've setup the Gemini API in Perplexica

But I have this setup (apparently the same on Perplexica) and no avail with the pipeline:

#An error occurred while searching Perplexica: 400 Client Error: Bad Request for url: http://192.168.1.14:3001/api/search 

You see anything wrong?

**Comment by semidark** [comment]
I created a revised version of the script that @ViffyGwaanl created.

It now supports streaming and has the references in a dropdown element, so they don't clutter the UI. I also added more error handling, since the original version of the script crashed my open-webui instance.

You can get it here: [perplexica-pipe-stream.py](https://gist.github.com/semidark/5566c988c67c5b28bf5b54afa0356422#file-perplexica-pipe-stream-py)
